### PR TITLE
Partial rewrite of r.valley.bottom

### PR DESCRIPTION
**Pull request by stevenpawley**
Hello Helmut,

The Multiresolution Index of Valley Bottom Flatness is an algorithm that I use very frequently as part of terrain and wetland predictive modelling. I noticed that the original GRASS version, whilst almost complete, was not producing the correct output (consisting of a grid of only 0s and 1s). Also due to the absence of cell-padding in GRASS GIS raster operations, the output would shrink quite substantially from the original DEM size. I know that the full range of parameters, including use of DEMs of differing resolutions, and access to changing the thresholds and shape parameters had not been implemented yet.

Please take a look at this rewritten version. The output, I think is correct now. Also, all of the parameter options are exposed and the complementary Multiresolution Index or Ridge Top Flatness represents an optional output. I also implemented a more fail-safe clean-up of temporary files, and I make sure the processing region is returned to the user's original region at the end of the script. The user can also specify any number of generalization steps using the 'levels' argument, because the calculation is now performed in a loop rather than each step separately. Additionally I manually grow the results of the various steps that represent moving windows to create the effect of cell padding so that the result doesn't shrink.

If you are happy with the changes, I can update the add ons svn.

Original algorithm's output:
image

New output:
image

### CODE REVIEW

File: r.valley.bottom.html [hunk 1]
```diff
@@ -1,40 +1,53 @@
...
+  
+  # calculate a shaded relief image
+  r.relief input=el_D782_6m@PERMANENT output=hs_D782_6m altitude=45 azimuth=315 zscale=4 scale=1                      
 
   # run <em>r.valley.bottom</em>
-  r.r.valley.bottom elevation=DEM
+  r.valley.bottom elevation=el_D782_6m@PERMANENT mrvbf=mrvbf_el_D782_6m t_slope=40
+  
+  # set colors
+  r.colors map=mrvbf_el_D782_6m color=bcyr -e
```
Comment (hellik) [inline r.valley.bottom.html]: tested your example; shouldn't it be 

`r.colors map=mrvbf_el_D782_6m color=bcyr -e **-n** `

to get the colored maps as in the screenshot?

**Comment by hellik** [comment]
Hi Steven,

great, thanks a lot! 

the addon script is tagged as experimental, as it was always more some kind of a proof of concept than a working script. I find the original paper quite confusing about all the equations and coarsening steps etc. unfortunately I couldn't find some time to improve the script in the last years.

I've tested your example, yes it seems to work quite well! very cool !!! the results looks plausible!

in the changes, I've seen you switched back to r.slope.aspect. this module shrinks the raster by one cell in every step; I wasn't aware of this for a very long time. a few years back Helena Mitasova suggested to use r.resamp.rst [1] instead: r64893 log: _r.slope.aspect -> r.resamp.rst slope (avoid region extent shrinking)_

have you already tried r.resamp.rst? if the resolution of the input raster is used in r.resamp.rst, no resampling is done, only slope is calculated, and no 1 pixel shrinking is done.

maybe there were also some other implications causing this shrinking in the svn version of the script.

in this PR, there is also a **.DS_Store** binary file. this file shouldn't go in a PR. could you modfy the PR?

I've left a small comment in the manual additions regarding r.color in the example. please have a look there.

let's think a little bit about the r.slope.aspect/r.resamp.rst options.

otherwise if you delete the binary file and have look in the manual example, I'll merge your PR and then you can update svn accordingly.

many thanks and kind regards
Helmut

[1] https://trac.osgeo.org/grass/browser/grass-addons/grass7/raster/r.valley.bottom/r.valley.bottom.py

**Comment by stevenpawley** [comment]
Thanks Helmut, yes it is quite a complicated algorithm and you did a great job in figuring it out in the first place. Regarding switching to r.slope.aspect, yes it shrinks the output raster by 1 cell and I could see why you used r.resample.rst. However, it is extremely slow for large grids (I work often work with 100,000 x 100,000 cell rasters) so it’s faster just to grow the map first, which is equivalent to cell padding. This applies not just to r.slope.aspect, but also to many GRASS GIS raster modules. You may noticed that I also grow the input before running r.resamp.interp because that will also shrink the output if using bilinear resampling, or more of using cubic. So does r.mapcalc if using a focal function, along with a slew of other GRASS tools. A general solution is needed for this like in SAGA GIS because although some users may be concerned about edge effects from using a cell padding method, I think most users expect the output raster to be the same size as the input.

Ah yes the silly .DS_store from occasionally using a Mac.... I’ll delete it!

**Comment by stevenpawley** [comment]
As a follow up, I might try and get in touch with the authors of the algorithm to see if they have a test dem and output from their original code. SAGA-GIS also has this algorithm and its very likely to be correct, but it would be nice to have an independent test because as you mentioned it is quite hard to interpret all of the steps based on the paper.

**Comment by stevenpawley** [comment]
Ok I've deleted the binary file. Apologies I couldn't see where your comment was on the manual page, but I have updated the example to provide a more complete example.

**Comment by stevenpawley** [comment]
Modified pull request

**Comment by hellik** [comment]
it seems I missed to push the button to send the comment regarding the manual example.

**Comment by hellik** [comment]
merged now your changes. many thanks! please go ahead to update the svn repo accordingly.

**Pull request merged**